zenoh-flow/src/types/message.rs

```diff
@@ -183,22 +314,169 @@ impl Message {
 }
 
 // Manual Ord implementation for message ordering when replay
-impl Ord for Message {
+impl Ord for LinkMessage {
     fn cmp(&self, other: &Self) -> Ordering {
         self.get_timestamp().cmp(&other.get_timestamp())
     }
 }
 
-impl PartialOrd for Message {
+impl PartialOrd for LinkMessage {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
         Some(self.cmp(other))
     }
 }
 
-impl PartialEq for Message {
+impl PartialEq for LinkMessage {
     fn eq(&self, other: &Self) -> bool {
         self.get_timestamp() == other.get_timestamp()
     }
 }
 
-impl Eq for Message {}
+impl Eq for LinkMessage {}
+
+/// A `Message<T>` is what is received on an `Input<T>`, typically after a call to `recv_async` or
+/// `recv`.
+///
+/// A `Message<T>` can either contain _Data_, `Data<T>` or signal a _Watermark_.
+pub enum Message<T: ZFData> {
+    Data(Data<T>),
+    Watermark,
+}
+
+/// A `Data<T>` is a convenience wrapper around `T`.
+///
+/// Upon reception, it transparently deserializes to `T` when the message is received serialized. It
+/// downcasts it to a `&T` when the data is passed "typed" through a channel.
+///
+/// ## Performance
+///
+/// When deserializing, an allocation is performed.
+#[derive(Debug)]
+pub struct Data<T> {
+    payload: Payload,
+    bytes: Option<Vec<u8>>,
+    // CAVEAT: Setting the value **typed** to `None` when it was already set to `Some` can cause a
+    // panic! The visibility of the entire structure is private to force us not to make such
+    // mistake.
+    // In short, tread carefully!
```

Comment:
Typo: treat

Reply:
Although "treat" would indeed work, I did want to use the expression "tread carefully" (= be cautious) 😛